### DISCUSSION TITLE
To get FS work with the new DBBC3 firmware DDC_E_v126

### DISCUSSION THREAD

**Discussion by vlbi-jun-yang** [nvi-inc/fs#187] [Q&A] [answered]
I have just noticed the release of new DBBC3 firmware, e.g. `DDC_E_v126.`  I am wondering how to get the existing FS work with the firmware. 

According to the manual, `DDC_E_v126` resembles `DDC_U_v125,` while supports less BBCs (only 8) and more bits (up to 8) quantinization. I guess that I can switch off the error message resulting from checking DBBC3 Firmware version via using `tnx` and then run the new firmware in a way as the same as `DDC_U_v125`. Right? 

Sure, the new firmware support can also be implemented in the next release of FS.

**Comment by ubach-ef** [comment]
I guess this will work just by changing the corresponding line in
/usr2/control/dbbc3.ctl
which provides a line to specify the number of BBCs and IFs.

**Comment by vlbi-jun-yang** [comment]
Thank Uwe for the comments. I have to correct my statement on 8-bit mode. After asking Sven, the firmware only supports one subband per Core3H board in 8-bit quantinization.

**Comment by wehimwich** [comment, marked as answer]
The FS currently only supports U and V firmware and v125 and v124 respectively of those. If E is close enough to one of those then the FS might be able to make support explicit, if perhaps only partial. I have no idea how similar the command sets are. Otherwise,  as you  say `tnx` is a way to shutoff error messages about the firmware version and you can use `dbbc3=....` (and perhaps `core3h=...`) to access features that the FS doesn't explicitly support.

Generally, you should only need to set the maximum number of BBCs that a DBBC3 can support in _dbbc3.ctl_. Using a subset should be okay. If some BBCs aren't present, the FS won't prevent attempts to use them, but presumably the DBBC3 will generate an error if you do. Likewise with IFs. However, the latter is a little trickier since it is assumed that the number of IFs is equal to the number of Core3H boards. This allows the FS to determine which boards aren't in use so that their output is disabled. If the IFs (and BBCs and Core3H boards) can vary with the firmware type and/or version number we may need some other way to specify these parameters.

The FS currently only supports one- and two-bit sampling.

**Comment by vlbi-jun-yang** [comment]
I have tried the method to stop displaying error messages. However, these messages were still logged in the log file. Therefore, it would be great to get FS to make a small patch and then support the sub-version firmware in the next release.

**Reply by wehimwich** [reply]
I think I understand. `tnx` was not really meant to handle an error that is so frequent and persistent. A rough calculation suggests the error messages would add about 1.7 MB to the log file in a 24-hour observation. Unfortunate, but compared to the multicast output, it is probably a small fraction (maybe 10%)

We are just finalizing _10.2_ and it does not have support for E_126. If it is small patch to support it, we can do it. From the little I know it will be more involved. If I have time before the final release, I can look into it. The real problem is the testing.

What we might be able to do instead is something like adding an environment variable to disable complaining about E from the multicast. That will get rid of the majority of the errors, but there would still be some (not much though) feedback of a wrong firmware version number from the setup procedure. It would also get rid of the need to disable the display of the errors for every restart of the FS. It is trade-off, better visibility for a firmware version error or a 10% increase in the log file. What is preferred? Maybe we could decrease the frequency of, but not eliminate, the errors from the multicast.

**Reply by ubach-ef** [reply]
E_126 is an U version with 8 BBCs only. Multicast and everything else appears to be the same as in the U_125 version. At least the FS is getting all the information that it is supposed to get. The main difference for the stations is that the E version with only 8 BBCs might have better filter shapes as the U version with 16 BBCs. As far as I know there are no new features/commands add or changes in any other behaviour involved.

**Reply by vlbi-jun-yang** [reply]
Onsala has a 2L2H DBBC3. You are welcome to use the DBBC3 backend to test your codes. Today, I am running the EVN experiment N23X1 with the DBBC3 backend. The test experiment will end in one hour. After that, it is free again. The next test experiment will be N23L1 on Mar 6. 

I have no preference on these temporary solutions. I am also happy to wait for the next release of 10.3 to well support the firmware.  The new feature of v126 is to support 8-bit quantisation for only one BBC sideband per Core3H board. However, there is no instruction for its usage for the time being. Hopefully, more information on its usage will be availalble in the next firmware release.

**Reply by wehimwich** [reply]
Then it _might_ be possible to get DDC_E v126 into the _10.2_; no promises. I would not be able to try until a later week. If I do, access for testing will be helpful. I will probably need to test with _V and _U too. There are questions:

- Since it is derived from _U, do I still need to send/read four bitmasks and two destinations? Even if not, it is easiest if it is backward compatible with that.
- Clock rate is 256?
- Are the Tsys display value reasonable or '$$$$'? 
- Time is shown in the Tsys display?
-  Any errors or other problems?

Maybe VGOS should be using _E.

BTW, I updated my estimates of the impact of the extra errors messages in the log for 24-hours to 1.7 MB and 10%. I had left out an error message before.

**Reply by vlbi-jun-yang** [reply]
As Uwe said, the firmware DDC_E_v126 is very similar to the firmware DDC_U_v126.  

> Then it _might_ be possible to get DDC_E v126 into the _10.2_; no promises. I would not be able to try until a later week. If I do, access for testing will be helpful. I will probably need to test with _V and _U too. 

I have not tried the firmware DDC _V yet. I will edit the configuration file and get the firmware work on our 2L2H DBBC3 today. 

> * Since it is derived from _U, do I still need to send/read four bitmasks and two destinations? Even if not, it is easiest if it is backward compatible with that.

Physiscally, only the first VSI bit mask and the first port Eth0 are useful. According to [the presentation](https://radiowiki.mpifr-bonn.mpg.de/lib/exe/fetch.php?media=na:sustainability:tog:2023_01:usage_of_the_dbbc3_firmware_and_software.pdf) by Sven Dornbusch during the TOG meeting, it is backward compatible.  It requires that all the 4 bit masks have the same bit mask. 

> * Clock rate is 256?

It is 128 MHz.

> * Are the Tsys display value reasonable or '$$$$'?

Yes, the Tsys data looked reasonable during the testing observations. 

> * Time is shown in the Tsys display?

Yes. It looks correct. See:
<img width="247" alt="image" src="https://user-images.githubusercontent.com/33714508/221917455-afdd97a5-899a-4787-bac5-06fe6008ae02.png">


> * Any errors or other problems?

No.

> Maybe VGOS should be using _E.

I hope so. In the observations N23X1, there are two frequency setups. Following your [online DBBC3 manual](https://nvi-inc.github.io/fs/releases/10/1/dbbc3_ops.html#_configuring_the_fs_for_use_with_a_dbbc3), the frequency setup was changed without reloading the firmware. 
 
> 
> BTW, I updated my estimates of the impact of the extra errors messages in the log for 24-hours to 1.7 MB and 10%. I had left out an error message before.

Thanks for the update.

**Reply by wehimwich** [reply]
That all sounds good. Thanks.

I didn't speak precisely about the sample rate. The sample rate that is commanded for DDC_U is `128`, but the _internal_ sample rate of `256` must be used to calculate the decimation.

Thank you for the pointer to the presentation. We'll have to see if we also get _V_125 and _U_128 supported too. Hopefully, they will come for "free" and  we only need changes for  _E_126.

**Reply by vlbi-jun-yang** [reply]
In the command line `vsi_samplerate`, the maximum sample rate for DDC_E is also 128 MHz.  This is a rather confusing change for DBBC2 users. The log files of these testing observations (N23C1 and N23X1) have also been uploaded to VLBEER. The two letter station code `OD` is used for the our DBBC3 backend.

**Reply by wehimwich** [reply]
@vlbi-jun-yang, from what I understand about the DDC_E (which is supposedly backward compatible with DDC_U), I don't see see anything anomalous in the logs related to the sample rate. I agree that it is odd that the command sample rate is `128` when the device uses `256` internally, but maybe I am missing the point of your comment.

**Comment by jfhquick** [comment]
On Thu, March 2, 2023 1:03 pm, Ed Himwich wrote:
> @vlbi-jun-yang, from what I understand about the DDC_E (which is supposedly
> backward compatible with DDC_U), I don't see see anything anomalous in the
> logs related to the sample rate. I agree that it is odd that the command
> sample rate is `128` when the device uses `256` internally, but maybe I am
> missing the point of your comment.

It's a matter of nomenclature.  You are commanding the VSI-H clock rate and
not the sampling rate.  The DDC_U/DDC_E always use two VSI-H connections in
parallel to carry the signal (with DDC_U using four in total) so the sampling
rate is then double the clock rate.

**Reply by wehimwich** [reply]
Thanks. It is a subtle point. The `128` is not the sample rate. It is the input clock rate even though it is set with a DBBC3 command named `vsi_samplerate`. To my surprise, the `help` page for `core3h_mode` command (**`help=core3h_mode`**) actually makes the distinction of it being the input clock rate. It does not explain the issue in detail though.

**Comment by wehimwich** [comment]
Closed by 46eace2c4a0961d962cc1f7bb0d27ac1b701aa8e
